### PR TITLE
WIP: Added -Y warning check options

### PR DESCRIPTION
**Pull request by mbeckerle**
These 4 were added.
    "-Ywarn-infer-any",
    "-Ywarn-inaccessible",
    "-Ywarn-nullary-unit",
    "-Ywarn-unused-import"

Unwise to merge this until we're at a stable point. 

Honestly the warn-nullary-unit change was a bunch of lines of code changed for basically no value. 
It doesn't require that the calls to the functions be changed to also have the parentheses.

I.e., it wants

`def init(): Unit = { ... }`

but you can still, without warning call this by just saying `init`. No parentheses. 

DAFFODIL-2145

### CODE REVIEW

File: daffodil-lib/src/main/scala/org/apache/daffodil/oolag/OOLAG.scala [hunk 1]
```diff
@@ -308,7 +308,7 @@ object OOLAG extends Logging {
      * Unconditionally, evaluate the LV arg in order to insure all checks for this
      * object are performed.
      */
-    private def requiredEvaluationsAlways(lv: OOLAGValueBase) : Unit = {
+    private def requiredEvaluationsAlways(lv: OOLAGValueBase) (): Unit = {
```
Comment (stevedlawrence) [inline daffodil-lib/src/main/scala/org/apache/daffodil/oolag/OOLAG.scala:311]: A few of the changes in this file look incorrect--looks like theres an extra set of parens. Is this a accidental find/replace due to the space before the colon?

File: daffodil-propgen/src/main/scala/org/apache/daffodil/propGen/PropertyGenerator.scala [hunk 2]
```diff
@@ -595,7 +595,7 @@ object Currency {
   def generateNonEnumStringInit(pgName: String, propInits: Seq[String]) = {
     val initFuncName = initialLowerCase(pgName)
     """
-  def """ + initFuncName + """Init() : Unit = {""" +
+  def """ + initFuncName + """Init() (): Unit = {""" +
```
Comment (stevedlawrence) [inline daffodil-propgen/src/main/scala/org/apache/daffodil/propGen/PropertyGenerator.scala:598]: This looks off too, I think this ends up generated an extra set of parens.

File: daffodil-propgen/src/test/scala/org/apache/daffodil/propGen/TestPropertyGenerator.scala [hunk 3]
```diff
@@ -65,7 +65,7 @@ class TestPropertyGenerator {
     val mx = pg.genAttributeGroup(sch)
     assertTrue(mx.contains(""" = convertToBoolean(findProperty("prefixIncludesPrefixLength").value)"""))
     assertTrue(mx.contains("""LengthKindMixin"""))
-    assertTrue(mx.contains("""def lengthPropertiesAGInit() : Unit = {"""))
+    assertTrue(mx.contains("""def lengthPropertiesAGInit() (): Unit = {"""))
```
Comment (stevedlawrence) [inline daffodil-propgen/src/test/scala/org/apache/daffodil/propGen/TestPropertyGenerator.scala:68]: Same issue here

**Review by stevedlawrence — APPROVED** [review]
+1, maybe a couple accidental find/replaces that aren't quiet right?

File: daffodil-propgen/src/test/scala/org/apache/daffodil/propGen/TestPropertyGenerator.scala (hunk 3 above)
Comment (mbeckerle) [reply daffodil-propgen/src/test/scala/org/apache/daffodil/propGen/TestPropertyGenerator.scala:68]: Yup. I'm counting on the CI to find the bugs for me.

Comment (stevedlawrence) [reply daffodil-propgen/src/test/scala/org/apache/daffodil/propGen/TestPropertyGenerator.scala:68]: Looks likethe CI passed, I guess this is technically valid syntax since scala allows multiple parameter lists for currying, and I guess it has no problem with an empty extra parameter list?

Comment (mbeckerle) [reply daffodil-propgen/src/test/scala/org/apache/daffodil/propGen/TestPropertyGenerator.scala:68]: Strange. But yes you can 
`def foo(): Unit = { .... }`
then invoke with just
`foo` not `foo()`
And scala supplies the implied added '()'. So it seems it can do this even for
`def foo()(): Unit = { .... }`
which invokes with 
`foo()`, again supplying the second added '()' implicitly. 

Grrrr. Feels like a bug, not a feature. 

I'll chase down all ") ()". 

I've been thinking about Scala 2.13 and Scala 3.0 eventually, but unless we cross-support them (like we did for 2.11 and 2.12 for a time) so that 2.12 continues to be supported, we're stuck because other important Scala libraries like Apache Spark are still on 2.12.

In other words, I suspect we have to turn on these other warnings that we need for code-quality again without waiting for new scala versions. I.e., in Scala 2.12, and cope with them best we can. E.g., I think the warn on unused and warn on dead-code are both pretty important things for the compiler to do for us. 

A wildcard worthy of investigation is whether the Java modules stuff (introduced in Java 9 I believe, and which seems not to be the OGSI stuff all over again) allows one to link together modules with different dependencies, so long as the crossing between them is a a Java API.

**Comment by jw3** [comment]
Using Scalafix makes some of these operations less painful... but some others more painful.

https://scalacenter.github.io/scalafix/

see #388 #393

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala [hunk 4]
```diff
@@ -386,7 +386,7 @@ final class UStateForSuspension(
   override def advanceAccessor: InfosetAccessor = die
   override def inspect: Boolean = die
   override def inspectAccessor: InfosetAccessor = die
-  override def fini: Unit = {
+  override def fini(): Unit = {
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:389]: This finalize method not being called suggests it is not needed. Could be dead code.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala [hunk 5]
```diff
@@ -530,7 +530,7 @@ final class UStateMain private (
   override def advanceAccessor: InfosetAccessor = inputter.advanceAccessor
   override def inspect: Boolean = inputter.inspect
   override def inspectAccessor: InfosetAccessor = inputter.inspectAccessor
-  override def fini: Unit = { inputter.fini }
+  override def fini(): Unit = { inputter.fini }
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:533]: ditto. Dead code?

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala [hunk 6]
```diff
@@ -1078,7 +1080,7 @@ trait DataOutputStreamImplMixin extends DataStreamCommonState
     setMaybeRelBitLimit0b(savedBitLimit0b, true)
   }
 
-  final override def validateFinalStreamState: Unit = {
+  final override def validateFinalStreamState(): Unit = {
```
Comment (mbeckerle) [inline daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala:1083]: This not being called indicates a bug or dead code.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DFDLDelimiter.scala [hunk 7]
```diff
@@ -445,7 +445,7 @@ class CharDelim(val char: Char, ignoreCase: Boolean) extends DelimBase {
   }
 
   lazy val typeName = "CharDelim"
-  def print = {
+  def print() = {
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DFDLDelimiter.scala:448]: Dead code - left over from debugging. Remove or comment out fully? Ditto for all other print() methods in this file.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DFDLDelimiter.scala [hunk 8]
```diff
@@ -412,7 +412,7 @@ class Delimiter {
 
 abstract class DelimBase extends Base {
   def typeName: String
-  def print: Unit
+  def print(): Unit
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DFDLDelimiter.scala:415]: Might be left over debug method. Removed?

**Review by tuxji — APPROVED** [review]
+1

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala (hunk 6 above)
Comment (mbeckerle) [reply daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala:1083]: TDML runner explicitly runs both regular and sax APIs.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala (hunk 4 above)
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:404]: Not dead.

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala (hunk 6 above)
Comment (mbeckerle) [reply daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala:1080]: Not dead code.

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DataStreamCommon.scala [hunk 9]
```diff
@@ -98,14 +98,4 @@ trait DataStreamCommon {
...
   def futureData(nBytesRequested: Int): ByteBuffer
-
-  /**
-   * Called once after each parse operation to verify final invariants for
-   * the implementation.
-   *
-   * Use to perform checks such as that data structures held in pools are
-   * all returned before end of parse.
-   */
-  def validateFinalStreamState: Unit
```
Comment (tuxji) [inline daffodil-io/src/main/scala/org/apache/daffodil/io/DataStreamCommon.scala:109]: Are you removing this code for a valid reason?

File: daffodil-io/src/main/scala/org/apache/daffodil/io/StringDataInputStreamForUnparse.scala [hunk 10]
```diff
@@ -84,5 +84,4 @@ final class StringDataInputStreamForUnparse
   override def skip(nBits: Long, finfo: FormatInfo): Boolean = doNotUse
   override def resetBitLimit0b(savedBitLimit0b: MaybeULong): Unit = doNotUse
   // $COVERAGE-ON$
-  override def validateFinalStreamState: Unit = {} // does nothing
```
Comment (tuxji) [inline daffodil-io/src/main/scala/org/apache/daffodil/io/StringDataInputStreamForUnparse.scala:87]: If this removal is valid. also remove the `// $COVERAGE-ON$`.

File: daffodil-lib/src/test/scala/passera/test/UnsignedCheck.scala [hunk 11]
```diff
@@ -30,14 +30,12 @@ import org.scalacheck._
...
 import org.scalacheck.Prop._
 import passera.unsigned._
-
 import org.junit.Test
 import org.junit.Assert._
 
 import scala.language.implicitConversions
 
 class UnsignedCheck {
-  import Gen._
```
Comment (tuxji) [inline daffodil-lib/src/test/scala/passera/test/UnsignedCheck.scala:40]: Removing an import should be safe as long as the code still compiles, but where is Gen defined?  I ran `rg -uu -w Gen` over my entire fully built checkout and it found only three lines, none of which appears to define Gen:

```
daffodil-lib/src/test/scala/passera/test/UnsignedCheck.scala
40:  import Gen._
52:  def genUInt: Gen[UInt] = for (n <- arbitrary[Int]) yield UInt(n)
61:  val nonNegLong = Gen.choose(0L, 0x00000000ffffffffL)
```

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala [hunk 12]
```diff
@@ -223,7 +223,7 @@ class DPathCompileInfo(
     with PreSerialization
     with HasSchemaFileLocation {
 
-  def initialize: Unit = {
+  lazy val initialize: Unit = {
     parents
   }
```
Comment (tuxji) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala:228]: Why this change, since I don't see any computation being done?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala [hunk 13]
```diff
@@ -397,34 +397,31 @@ final class UStateForSuspension(
...
+  // $COVERAGE-OFF$
   override def advance: Boolean = die
   override def advanceAccessor: InfosetAccessor = die
   override def inspect: Boolean = die
   override def inspectAccessor: InfosetAccessor = die
-  override def fini: Unit = {
-    //do nothing
-  }
+  override def fini(): Unit = die
   override def inspectOrError = die
```
Comment (tuxji) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:406]: Why does `fini()` get parentheses but `inspectOrError` doesn't get parentheses?

**Review by tuxji — CHANGES_REQUESTED** [review]
I noticed you asked for another review and I saw some old code was carried forward in build.sbt, so you need to fix that file.

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala [hunk 14]
```diff
@@ -1081,10 +1076,6 @@ trait DataOutputStreamImplMixin extends DataStreamCommonState
     setMaybeRelBitLimit0b(savedBitLimit0b, true)
   }
 
-  final override def validateFinalStreamState: Unit = {
-    // nothing to validate
-  }
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala:1086]: Dead code. Base is never called.

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DataStreamCommon.scala (hunk 9 above)
Comment (mbeckerle) [reply daffodil-io/src/main/scala/org/apache/daffodil/io/DataStreamCommon.scala:109]: Dead code. Base is never called.

File: daffodil-io/src/main/scala/org/apache/daffodil/io/StringDataInputStreamForUnparse.scala (hunk 10 above)
Comment (mbeckerle) [reply daffodil-io/src/main/scala/org/apache/daffodil/io/StringDataInputStreamForUnparse.scala:87]: dead code. Will fix.

File: daffodil-lib/src/test/scala/passera/test/UnsignedCheck.scala (hunk 11 above)
Comment (mbeckerle) [reply daffodil-lib/src/test/scala/passera/test/UnsignedCheck.scala:40]: IntelliJ removed this automatically I assume. 
Gen is defined in scalacheck/scalacheck_2.12/scalacheck_2.12-1.15.4-sources.jar!/org/scalacheck/Gen.scala

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala (hunk 13 above)
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:406]: Yeah. I think this should be without the (). Clearly scala is not warning about this. It was warning about it at one point.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala (hunk 12 above)
Comment (tuxji) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala:228]: I found a StackOverflow [answer](https://stackoverflow.com/questions/7600910/difference-between-function-with-parentheses-and-without) which said that calling nullary methods without an explicit () is no longer optional in Scala 2.13 or 3.0.  BTW, the change wasn't made to annoy people; it was a necessary side effect of another change to improve type inference by allowing eta expansion of Scala functions.  What this tells me is that you shouldn't bother changing `initialize` from a def to a lazy val.  When we upgrade to 2.13, then scalac will enforce consistency between calls and definitions and we can fix any remaining calls without () at that time.

Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala:228]: This is super helpful. So use () when the method has side effects, but no args. Leave out parens if it is side-effect free. 

Note that a lazy val that only reads things is still considered side-effect free, even though it has this lazy-val once-only behavior. So if you define a def, intending to possibly override with a lazy val, then no parameters is correct, as a lazy val can't have even an empty argument list. 

Also, **_conceptual_** side-effect free applies. Initializing something that happens once as part of setup of some sort is NOT considered a side effect. 

A side-effect is something that has to happen *every* time the method is used. E.g, like assigning something each time, or doing an I/O operation each time.

Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala:228]: Actually, based on this. We really can't use "-Ywarn-nullary-unit", because it forces us to use () for def that want to be overridden as lazy val.

Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala]: This code is back to being a def with no args.

File: build.sbt [hunk 15]
```diff
@@ -157,6 +157,25 @@ lazy val commonSettings = Seq(
...
+    "-Ywarn-infer-any",
+    "-Ywarn-inaccessible",
+    "-Ywarn-nullary-unit",
+    "-Ywarn-unused-import"
+  ),
+  // Workaround issue that some options are valid for javac, not javadoc.
+  // These javacOptions are for code compilation only. (Issue sbt/sbt#355)
+  Compile / compile / javacOptions  ++= Seq(
+    "-Werror",
+    "-Xlint:deprecation"
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply build.sbt]: fixed

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala [hunk 16]
```diff
@@ -397,34 +397,31 @@ final class UStateForSuspension(
...
-
   override def pushDelimiters(node: DelimiterStackUnparseNode) = die
   override def popDelimiters() = die
+  // $COVERAGE-ON$
+
+  override def groupPos = 0 // was die, but this is called when copying state during debugging
+  override def currentInfosetNodeMaybe = Maybe(currentInfosetNode)
+  override def arrayPos = occursIndex
+  override def childPos = 0 // was die, but this is called when copying state during debugging.
+
```
Comment (tuxji) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:424]: I looked at why groupPos and childPos's lines (420, 423) still aren't covered by tests (their original locations were inside the $COVERAGE-OFF$ block).  My best guess is that we don't have a test forcing the interactive debugger to copy unparser state.  I'll still give you a +1.

**Review by tuxji — APPROVED** [review]
+1
